src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs

```diff
@@ -0,0 +1,1200 @@
+                State = State.Stopped;
+                return result;
+            }
+        }
+
+        int _fileDescriptor;
+        GCHandle _handle;
+        Queue<TransferOperation> _receiveQueue;
+        Queue<TransferOperation> _sendQueue;
+        Queue<AcceptOrConnectOperation> _acceptOrConnectQueue;
+        object _closeLock = new object();
+        object _queueLock = new object();
+        SocketAsyncEngine _engine;
+        uint _registeredEvents;
+
+        public SocketAsyncContext(int fileDescriptor, SocketAsyncEngine engine)
+        {
+            _fileDescriptor = fileDescriptor;
+            _engine = engine;
+
+            var events = Interop.libc.EPOLLIN | Interop.libc.EPOLLOUT | Interop.libc.EPOLLRDHUP | Interop.libc.EPOLLET;
+
+            Interop.Error errorCode;
+            if (!_engine.TryRegister(this, _fileDescriptor, events, 0, ref _handle, out errorCode))
+            {
+                throw new Exception(string.Format("SocketAsyncContext: {0}", errorCode));
+            }
+
+            _registeredEvents = events;
+        }
+
+        public void Close()
+        {
+            lock (_closeLock)
+            lock (_queueLock)
+            {
+                // Force a HUP event in order to drain the queues.
+                HandleEvents(Interop.libc.EPOLLHUP);
+            }
+        }
+
+        private bool TryBeginOperation<TOperation>(ref Queue<TOperation> queue, TOperation operation, out bool isStopped)
+            where TOperation : AsyncOperation
+        {
+            lock (_queueLock)
+            {
+                switch (queue.State)
+                {
+                    case State.Stopped:
+                        isStopped = true;
+                        return false;
+
+                    case State.Clear:
+                        break;
+
+                    case State.Set:
+                        isStopped = false;
+                        queue.State = State.Clear;
+                        return false;
+                }
+
+                queue.Enqueue(operation);
+                isStopped = false;
+                return true;
+            }
+        }
+
+        private void EndOperation<TOperation>(ref Queue<TOperation> queue)
+            where TOperation : AsyncOperation
+        {
+            lock (_queueLock)
+            {
+                Debug.Assert(!queue.IsStopped);
+
+                queue.Dequeue();
+            }
+        }
+
+        public SocketError AcceptAsync(byte[] socketAddress, int socketAddressLen, Action<int, byte[], int, SocketError> callback)
+        {
+            Debug.Assert(callback != null);
+
+            int acceptedFd;
+            SocketError errorCode;
+            if (TryCompleteAccept(_fileDescriptor, socketAddress, ref socketAddressLen, out acceptedFd, out errorCode))
+            {
+                ThreadPool.QueueUserWorkItem(_ => callback(acceptedFd, socketAddress, socketAddressLen, errorCode), null);
```

Comment:
As written, this is going to cause a closure allocation even if TryCompleteAccept returns false, as it's capturing state outside of the scope of the if block.  It's also going to cause a delegate allocation when invoked.  You could address both of those by changing it to something like:
```
ThreadPool.QueueUserWorkItem(s =>
{
    var args = (Tuple<...>)s;
    args.Item1(args.Item2, args.Item3, args.Item4, args.Item5);
}, Tuple.Create(callback, acceptedFd, socketAddress, socketAddressLen, errorCode));
```